packages/@icon-magic/icon-models/src/icon.ts

```diff
@@ -0,0 +1,176 @@
+import * as debugGenerator from 'debug';
+import * as fs from 'fs-extra';
+import * as path from 'path';
+
+import { Asset } from './asset';
+import { Flavor } from './flavor';
+import {
+  AssetResolution,
+  AssetSize,
+  BuildConfig,
+  DistributeConfig,
+  FlavorConfig,
+  GenerateConfig,
+  IconConfig
+} from './interface';
+import { exists, isTypeSVG } from './utils/files';
+
+/**
+ * Encapsulates what an Icon means It is referenced by a path to it's directory
+ * and the directory of icons can be assumed to contain all the different
+ * variants and flavors in which the icon is available, in all of it's different
+ * types. The config itself is generally more concise and human readable but this
+ * class supplements it by providing methods on it and filling in the gaps where
+ * they don't exist
+ */
+export class Icon {
+  iconPath: string;
+  variants: Asset[];
+  sourceConfigFile: string;
+  sizes: AssetSize[];
+  resolutions: AssetResolution[];
+  iconName: string;
+  flavors: Map<string, Flavor>;
+  outputPath: string;
+  build?: BuildConfig;
+  generate?: GenerateConfig;
+  distribute?: DistributeConfig;
+  private debug: debugGenerator.IDebugger;
+
+  /**
+   * Creates an Icon instance by creating sub classes for it's variants and
+   * flavors
+   * @param config Config read from a config file containing all the icon's
+   * information
+   */
+  constructor(config: IconConfig) {
+    this.debug = debugGenerator('icon-magic:icon-models:icon');
+    // copy over all the properties in the config
+    Object.assign(this, config);
```

Comment:
Because IconConfig has a well defined interface, we can get better type safety by doing:
```
for (let key of Object.keys(config)) {
  this[key] = config[key];
}
```
And we can enable that ensure instantiation rule in the tsconfig. This will help us in the long run.